chef/cookbooks/nova/recipes/ceph.rb

```diff
@@ -0,0 +1,185 @@
+#
+# Copyright (c) 2015 SUSE Linux GmbH.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Cookbook Name:: nova
+# Recipe:: ceph
+#
+
+has_internal = false
+has_external = false
+
+cinder_controller = search(:node, "roles:cinder-controller")
+return unless cinder_controller.length > 0
+
+# First loop to find if we have internal/external cluster
+cinder_controller[0][:cinder][:volumes].each_with_index do |volume, volid|
+  next unless volume[:backend_driver] == "rbd"
+
+  has_internal ||= true if volume[:rbd][:use_crowbar]
+  has_external ||= true unless volume[:rbd][:use_crowbar]
+end
+
+if has_internal
+  ceph_env_filter = " AND ceph_config_environment:ceph-config-default"
+  ceph_servers = search(:node, "roles:ceph-osd#{ceph_env_filter}") || []
+  if ceph_servers.length > 0
+    include_recipe "ceph::keyring"
+  else
+    message = "Ceph was not deployed with Crowbar yet!"
+    Chef::Log.fatal(message)
+    raise message
+  end
+end
+
+if has_external
+  # Ensure ceph is available here
+  if node[:platform] == "suse"
+    # install package in compile phase because we will run "ceph -s"
+    package "ceph-common" do
+      action :nothing
+    end.run_action(:install)
+  end
+end
+
+package "qemu-block-rbd" do
+  action :install
+  only_if { node[:platform_version].to_f >= 12.0 }
+end
+
+# Second loop to do our setup
+cinder_controller[0][:cinder][:volumes].each_with_index do |volume, volid|
+  next unless volume[:backend_driver] == "rbd"
+
+  if volume[:rbd][:use_crowbar]
+    ceph_conf = "/etc/ceph/ceph.conf"
+    admin_keyring = "/etc/ceph/ceph.client.admin.keyring"
+  else
+    ceph_conf = volume[:rbd][:config_file]
+    admin_keyring = volume[:rbd][:admin_keyring]
+
+    if ceph_conf.empty? || !File.exists?(ceph_conf)
+      Chef::Log.info("Ceph configuration file is missing; skipping the ceph setup for backend #{volume[:backend_name]}")
+      next
+    end
+
+    if !admin_keyring.empty? && File.exists?(admin_keyring)
+      Chef::Log.info("Using external ceph cluster for nova #{volume[:backend_name]} backend, with automatic setup.")
+    else
+      Chef::Log.info("Using external ceph cluster for nova #{volume[:backend_name]} backend, with no automatic setup.")
+      next
```

Comment:
It would be reasonable to use empheral ceph backend only for internal ceph integration, because we will be sure that we will have access to admin keyring.